onnx/defs/tensor_proto_util.cc

```diff
@@ -30,45 +31,63 @@ namespace ONNX_NAMESPACE {
-          *end_byte = temp;                                                \
-          ++start_byte;                                                    \
-          --end_byte;                                                      \
-        }                                                                  \
-      }                                                                    \
-    }                                                                      \
-    /* raw_data.c_str()/bytes is a byte array and may not be properly  */  \
-    /* aligned for the underlying type */                                  \
-    /* We need to copy the raw_data.c_str()/bytes as byte instead of  */   \
-    /* copying as the underlying type, otherwise we may hit memory   */    \
-    /* misalignment issues on certain platforms, such as arm32-v7a */      \
-    const size_t raw_data_size = raw_data.size();                          \
-    res.resize(raw_data_size / sizeof(type));                              \
-    memcpy(reinterpret_cast<char*>(res.data()), bytes, raw_data_size);     \
-    return res;                                                            \
+#define DEFINE_PARSE_DATA(type, typed_data_fetch)                              \
+  template <>                                                                  \
+  const std::vector<type> ParseData(const TensorProto* tensor_proto) {         \
+    std::vector<type> res;                                                     \
+    if (!tensor_proto->has_data_type() ||                                      \
+      tensor_proto->data_type() == TensorProto_DataType_UNDEFINED) {           \
+      fail_shape_inference("The type of Tensor: ", tensor_proto->name(),       \
+        " is undefined so it cannot be parsed.");                              \
+    } else if (tensor_proto->has_data_location() &&                            \
+      tensor_proto->data_location() == TensorProto_DataLocation_EXTERNAL) {    \
+      fail_shape_inference("Cannot parse data from external tensors. Please ", \
+        "load external data into raw data for tensor: ", tensor_proto->name());\
+    } else if (!tensor_proto->has_raw_data()) {                                \
+      const auto& data = tensor_proto->typed_data_fetch();                     \
+      int expected_size = 1;                                                   \
+      for (int i = 0; i < tensor_proto->dims_size(); ++i) {                    \
+        expected_size *= tensor_proto->dims(i);                                \
+      }                                                                        \
+      if (tensor_proto->dims_size() != 0 && data.size() != expected_size) {    \
+        fail_shape_inference("The data size of tensor: ", tensor_proto->name(),\
+        " was expected to have (calculated by dims) ", expected_size,          \
+        " but the actual data size was ", data.size());                        \
```

Comment:
rephrase error message as "Data size mismatch. Tensor: <name> expected size <expected_size> does not match the actual size <data.size()>"